### PR TITLE
new actions, code style consistency, async await, disabled actions and bugfixes

### PR DESCRIPTION
**Pull request by dbuezas**
The options html is now prettier. Hope you like it :)

### CODE REVIEW

**Comment by dbuezas** [comment]
screen shot 2017-08-05 at 19 17 48

**Comment by dbuezas** [comment]
everything according to existing style :)
I also replaced all double quotes to single quotes and ported all actions to await async. That saves 65 lines of code, a lot of columns, reduces nesting and improves readability.
the only fat arrow functions now are one liners inside .filter or .map. Those return a value so there it is worth it :)
Hope you like it. I love this extension by the way. These new javascript syntax sugar could also attract more people to add actions, since I think they are now way easier to understand and follow.

BTW, the fuzzy search module seems to sort in a sort of wacky way, it doesn't care if you have a full word match.

**Comment by ssundarraj** [comment]
> Hope you like it. I love this extension by the way. These new javascript syntax sugar could also attract more people to add actions, since I think they are now way easier to understand and follow.

Thanks. This is great! I should probably get around to porting the rest of the syntax to use ES6 some time soon :)
> BTW, the fuzzy search module seems to sort in a sort of wacky way, it doesn't care if you have a full word match.

That's odd. Can you please file a bug with a screenshot if you have one handy?

**Comment by ssundarraj** [comment]
This is what I get when I build the extension and test it. Seems broken to me. Maybe you're missing some changes?

screen shot 2017-08-06 at 12 33 29 am

**Comment by dbuezas** [comment]
> Thanks. This is great! I should probably get around to porting the rest of the syntax to use ES6 some time soon :)

Cool! I can definitely recommend to use es-lint, there are some non declared variables that may lead to bugs and linting tells you about that :). Also it would help defining a coding style and having everybody respect it, plus it also gives recommendations about best practices, etc. Can help setting that up if you wish.
It would also be great to require other files instead of using globals, but at the current size of the project it is still manageable.

> That's odd. Can you please file a bug with a screenshot if you have one handy?

screen shot 2017-08-06 at 11 31 09
in this case I would expect the algorithm to prioritise `Switch to: [....] Youtube`

**Comment by dbuezas** [comment]
The bug you found is fixed now 🎉 
I accidentally changed the format in which custom actions were stored and didn't notice it myself because I "fixed" it in the `populateSuggestionList` function instead of in the `addCommand`function 😅 

While I was debugging I got stuck because the stored custom action's code wasn't valid, so I added a couple of small features:

1. Don't store custom actions that cannot be 'eval'ed.
screen shot 2017-08-06 at 11 41 03

2. Show an error message in the popup if an action throws:
screen shot 2017-08-06 at 11 29 26
It also points out to open the popup's DevTools and console.errors the function and the exception so users don't get lost while debugging. This also applies to default actions, so it should help extension devs also (I got lost a bit at first).
screen shot 2017-08-06 at 11 51 48

3. I added a link to the chrome api to the user custom actions comments:
screen shot 2017-08-06 at 11 49 25
I wished I had that at hand yesterday when I started adding the "reopen closed tab" action :)


4. Open Commander Options from the popup:
screen shot 2017-08-06 at 11 50 23

**Comment by dbuezas** [comment]
the new fuzzy library works great! 
I simplified stuff a bit and reduced code size another 20 lines in main.js.
also now searches can be added in the actions.js like this:
```{
    text: 'Define in Dictionary.com',
    action: triggerSearch('def'),
    triggers: ['def', 'define', 'dictionary'],
    queryToUrl: (q) => `http://dictionary.reference.com/browse/${q}`,
  }
```
and can also be disabled in the options since they are handled like normal actions now

**Comment by dbuezas** [comment]
I have a gizillion tabs open and switch youtube now only gives the correct result 🎉 
screen shot 2017-08-06 at 14 26 35

**Comment by ssundarraj** [comment]
The bug still exists for me. I suspect it's because some of the changes aren't compatible with the grunt config. When you use the extension are you using the distribution version of it?

If not, can you run `grunt` and load the extension from the 'dist' directory and see if it works? 

screen shot 2017-08-07 at 11 31 55 pm

**Comment by ssundarraj** [comment]
I think you'll need to update L29 of the gruntfile.

**Comment by dbuezas** [comment]
yep, never tried to build the extension 😅.
This should fix it.
I added a simple regex to replace the css to their .min.css and another one to replace everything between
`<!-- <jsminify dest="some.file.name"> -->`
and
`<!-- </jsminify> -->`
with a script tag with src = `<whatever was in dest>`

one still has to add the required js files in the Gruntfile.js but at least the replace term is easier (I think)

File: Gruntfile.js [hunk 1]
```diff
@@ -24,10 +34,16 @@ module.exports = function(grunt) {
         options: {
             processContentExclude: ['**/*.{png,gif,jpg,ico,psd}'],
             process: function (content, srcpath){
-            content = content.replace("css/options.css", "options.min.css");
-            content = content.replace("js/options.js", "options.min.js");
-            return content.replace("<script src=\"js/vendor/fuse.min.js\"></script>\n    <script src=\"js/actions.js\"></script>\n    <script src=\"js/main.js\"></script>", "<script src=\"built.min.js\"></script>").replace("css/main.css", "main.min.css");
-          }
+                return content
```
Comment (ssundarraj) [inline Gruntfile.js:37]: Indentation looks off here

File: src/js/actions.js [hunk 2]
```diff
@@ -1,312 +1,262 @@
...
-}
-
-function searchDictionary(query) {
-    return function() {
-        query = encodeURI(query);
-        chrome.tabs.create({url: "http://dictionary.reference.com/browse/" + query});
-    }
-}
-
+chromeP = new ChromePromise();
```
Comment (ssundarraj) [inline src/js/actions.js]: Please rename chromeP -> chromePromise

File: src/js/main.js [hunk 3]
```diff
@@ -1,122 +1,111 @@
...
-        var newSuggestion = highlightedSuggestion.nextSibling;
-        if (!newSuggestion){
-            var allSuggestions = document.getElementsByClassName("suggestion");
-            newSuggestion = allSuggestions[allSuggestions.length - 1]
-        }
-        changeHighlighted(newSuggestion);
-        reScroll();
-        return false;
+  switch (e.which){
+    case (40):{// down
```
Comment (ssundarraj) [inline src/js/main.js]: nit: add space before the comment

File: src/js/main.js [hunk 4]
```diff
@@ -125,42 +114,75 @@ function handleMouseover(e){
...
+        var suggestionTag = document.createElement('li');
+        suggestionTag.className = 'suggestion';
         suggestionTag.innerHTML = escapeHtml(suggestion.text);
-        suggestionTag.onclick = suggestion.action;
+        suggestionTag.onclick = async function() {
+          try {
+            await suggestion.action();
+          } catch (e) {
+            document.body.innerHTML = (
+`Error executing action [${escapeHtml(suggestion.text)}]:
```
Comment (ssundarraj) [inline src/js/main.js]: Can this be indented?

**Review by ssundarraj — COMMENTED** [review]
Looks good minus some minor changes. Thanks!

Can you also clean up the commit log so it looks neater? It would be great if you can squash commits that are part of the same function into a single commit but if that's not possible, I'd settle for one monolithic commit with better descriptions.

File: Gruntfile.js (hunk 1 above)
Comment (dbuezas) [reply Gruntfile.js:37]: it was off before, old lines 27 to 29 are inside the process function.
regarding the `.replace` lines, I'd rather have them with the same identation as the return statement, what is your preference?

File: src/js/main.js (hunk 4 above)
Comment (dbuezas) [reply src/js/main.js]: you're right, it is HTML so it doesn't matter

File: Gruntfile.js (hunk 1 above)
Comment (dbuezas) [reply Gruntfile.js:37]: optn1:
```
process: function (content, srcpath){
    return content
      .replace(
        /href="css\/([^"]+?)\.css"/g,
        'href="$1.min.css"'
      )
      .replace(
        /<!-- <jsminify dest="(.+)"> -->[\S\s]+?<!-- <\/jsminify> -->/g,
        '<script src="$1"></script>'
      );
}
```
optn2:
```
process: function (content, srcpath){
    return content
    .replace(
      /href="css\/([^"]+?)\.css"/g,
      'href="$1.min.css"'
    )
    .replace(
      /<!-- <jsminify dest="(.+)"> -->[\S\s]+?<!-- <\/jsminify> -->/g,
      '<script src="$1"></script>'
    );
}
```
optn3:
```
process: function (content, srcpath){
    return content.replace(
      /href="css\/([^"]+?)\.css"/g,
      'href="$1.min.css"'
    ).replace(
      /<!-- <jsminify dest="(.+)"> -->[\S\s]+?<!-- <\/jsminify> -->/g,
      '<script src="$1"></script>'
    );
}
```

**Comment by dbuezas** [comment]
this took way too much effort but commits are now squashed and I learned some more git stuff :P

File: Gruntfile.js (hunk 1 above)
Comment (ssundarraj) [reply Gruntfile.js:37]: Oh my bad. Please leave it as is.

**Comment by ssundarraj** [comment]
Thanks David!

**Comment by dbuezas** [comment]
🎉 :)

**Comment by dbuezas** [comment]
next PRs should be more orderly

**Pull request merged**